lib/bolt/transport/orch/connection.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+module Bolt
+  module Transport
+    class Orch < Base
+      class Connection
+        attr_reader :logger, :key
+
+        CONTEXT_KEYS = Set.new(%i[plan_name description params]).freeze
+
+        def self.get_key(opts)
+          [
+            opts['service-url'],
+            opts['task-environment'],
+            opts['token-file']
+          ].join('-')
+        end
+
+        def initialize(opts, plan_context, logger)
+          @logger = logger
+          @key = self.class.get_key(opts)
+          client_keys = %w[service-url token-file cacert]
+          client_opts = client_keys.each_with_object({}) do |k, acc|
+            acc[k] = opts[k] if opts.include?(k)
+          end
+          logger.debug("Creating orchestrator client for #{client_opts}")
+
+          @client = OrchestratorClient.new(client_opts, true)
+          @plan_job = start_plan(plan_context)
+
+          @environment = opts["task-environment"]
+        end
+
+        def start_plan(plan_context)
+          if plan_context
+            begin
+              opts = plan_context.select { |k, _| CONTEXT_KEYS.include? k }
+              @client.command.plan_start(opts)['name']
+            rescue OrchestratorClient::ApiError => e
+              if e.code == '404'
+                @logger.debug("Orchestrator #{key} does not support plans")
+              else
+                @logger.error("Failed to start a plan with orchestrator #{key}: #{e.message}")
+              end
+              nil
+            end
+          end
+        end
+
+        def finish_plan(plan_result)
+          if @plan_job
+            @client.command.plan_finish(
+              result: plan_result.value,
```

Comment:
Need to pass `plan_job` here. Also this fails if result is nil.

Reply:
Hmm, if I fix those two issues locally, I get `Failed to finish plan on -production-: Some tasks for this plan job have not finished running`

Reply:
yeah me too, I'm looking into whats wrong with orch there.

regardless I think the solution for bolt is to not finish plans until irving z